pkg/kubectl/cmd/util/openapi/openapi_cache.go

```diff
@@ -0,0 +1,193 @@
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/kubernetes/pkg/version"
+)
+
+func init() {
+	registerBinaryEncodingTypes()
+}
+
+const openapiFileName = "openapi_cache"
+
+type cachingOpenAPIClient struct {
+	version      string
+	client       discovery.OpenAPISchemaInterface
+	cacheDirName string
+}
+
+// newCachingOpenAPIClient returns a new discovery.OpenAPISchemaInterface
+// that will read the openapi spec from a local cache if it exists, and
+// if not fetch an openapi spec using client
+// client: used to fetch a new openapi spec if a local cache is not found
+// version: the server version and used as part of the cache file location
+// cacheDir: the directory under which the cache file will be written
+func newCachingOpenAPIClient(client discovery.OpenAPISchemaInterface, version, cacheDir string) *cachingOpenAPIClient {
+	return &cachingOpenAPIClient{
+		client:       client,
+		version:      version,
+		cacheDirName: cacheDir,
+	}
+}
+
+// openAPIData returns an openapi spec.
+// It will first attempt to read the spec from a local cache
+// If it cannot read a local cache, it will read the file
+// using the client and then write the cache.
+func (c *cachingOpenAPIClient) openAPIData() (*Resources, error) {
+	// Try to use the cached version
+	if c.useCache() {
+		doc, err := c.readOpenAPICache()
+		if err == nil {
+			return doc, nil
+		}
+	}
+
+	// No cached version found, download from server
+	s, err := c.client.OpenAPISchema()
+	if err != nil {
+		glog.Warningf("Failed to download openapi data %v", err)
+		return nil, err
+	}
+
+	oa, err := newOpenAPIData(s)
+	if err != nil {
+		glog.Warningf("Failed to parse openapi data %v", err)
+		return nil, err
+	}
+
+	// Try to cache the openapi spec
+	if c.useCache() {
+		err = c.writeToCache(oa)
+		if err != nil {
+			// Just log an warning, no need to fail the command since we got the data we need
+			glog.Warningf("Unable to cache openapi spec %v", err)
+		}
+	}
+
+	// Return the parsed data
+	return oa, nil
+}
+
+// useCache returns true if the client should try to use the cache file
+func (c *cachingOpenAPIClient) useCache() bool {
+	return len(c.version) > 0 && len(c.cacheDirName) > 0
+}
+
+// readOpenAPICache tries to read the openapi spec from the local file cache
+func (c *cachingOpenAPIClient) readOpenAPICache() (*Resources, error) {
+	// Get the filename to read
+	filename := c.openAPICacheFilename()
+
+	// Read the cached file
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode the openapi spec
+	s, err := c.decodeSpec(data)
+
+	return s, err
+}
+
+// decodeSpec binary decodes the openapi spec
+func (c *cachingOpenAPIClient) decodeSpec(data []byte) (*Resources, error) {
+	b := bytes.NewBuffer(data)
+	d := gob.NewDecoder(b)
+	parsed := &Resources{}
+	err := d.Decode(parsed)
+	return parsed, err
+}
+
+// encodeSpec binary encodes the openapi spec
+func (c *cachingOpenAPIClient) encodeSpec(parsed *Resources) ([]byte, error) {
```

Comment:
No.  :( As discussed offline, the Ref type which contains the json reference has private fields that are dropped by binary serialization.